File: leetcode-04/solutions/python3/0127.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        set_ = set(wordList)

        if endWord not in set_:
            return 0

        n = len(beginWord)

        ans = 0

        queue1 = set([beginWord])
        queue2 = set([endWord])

        while queue1 and queue2:
            ans += 1
            if len(queue1) > len(queue2):
                queue1, queue2 = queue2, queue1
            queue = set()
            for word in queue1:
                for i in range(n):
                    for j in string.ascii_lowercase:
                        newWord = word[:i] + j + word[i + 1:]
                        if newWord in queue2:
                            return ans + 1
                        if newWord not in set_:
                            continue
                        set_.remove(newWord)
                        queue.add(newWord)
            queue, queue1 = queue1, queue

        return 0

File: leetcode-04/solutions/cpp/0127.cpp
class Solution {
 public:
  int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
    unordered_set<string> set(wordList.begin(), wordList.end());

    if (!set.count(endWord)) return 0;

    const int n = beginWord.length();

    int ans = 0;

    unordered_set<string> queue1{beginWord};
    unordered_set<string> queue2{endWord};

    while (!queue1.empty() && !queue2.empty()) {
      ans++;
      if (queue1.size() > queue2.size()) swap(queue1, queue2);
      unordered_set<string> queue;
      for (string word : queue1) {
        for (int i = 0; i < n; i++) {
          char c = word[i];
          for (int j = 'a'; j <= 'z'; j++) {
            word[i] = j;
            if (queue2.count(word)) return ans + 1;
            if (!set.count(word)) continue;
            set.erase(word);
            queue.insert(word);
          }
          word[i] = c;
        }
      }
      swap(queue, queue1);
    }

    return 0;
  }
};